sabnzbd/articlecache.py

```diff
@@ -22,76 +22,142 @@
+    MEBI,
+    GIGI,
+    ANFO,
+    ARTICLE_CACHE_UPPER_PERCENTAGE,
+    ARTICLE_CACHE_LOWER_PERCENTAGE,
+)
+from sabnzbd.nzb import Article, NzbFile, NZO_LOCK
+from sabnzbd.misc import to_units
 
 # Operations on the article table are handled via try/except.
 # The counters need to be made atomic to ensure consistency.
 ARTICLE_COUNTER_LOCK = threading.RLock()
 
+_SECONDS_BETWEEN_FLUSHES = 0.5
 
-class ArticleCache:
+
+class ArticleCache(threading.Thread):
     def __init__(self):
+        super().__init__()
+        self.shutdown = False
         self.__cache_limit_org = 0
         self.__cache_limit = 0
         self.__cache_size = 0
-        self.__article_table: dict[Article, bytes] = {}  # Dict of buffered articles
-
-        self.assembler_write_trigger: int = 1
+        self.__article_table: dict[Article, bytearray] = {}  # Dict of buffered articles
+        self.__cache_size_cv: threading.Condition = threading.Condition(ARTICLE_COUNTER_LOCK)
+        self.__next_flush: float = 0
+        self.__cache_size_upper: int = 0  # Force flush trigger
+        self.__cache_size_lower: int = 0  # Resume normal behaviour once cache recovers after force flush
 
         # On 32 bit we only allow the user to set 1GB
         # For 64 bit we allow up to 4GB, in case somebody wants that
         self.__cache_upper_limit = GIGI
         if sabnzbd.MACOS or sabnzbd.WINDOWS or (struct.calcsize("P") * 8) == 64:
             self.__cache_upper_limit = 4 * GIGI
 
+    def stop(self):
+        self.shutdown = True
+        with self.__cache_size_cv:
+            self.__cache_size_cv.notify_all()
+
+    def __needs_flush(self) -> bool:
+        """
+        Should we flush the cache?
+        Only if direct write is supported and cache usage is over the upper limit.
+        Or the downloader is paused and cache is not empty.
+        """
+        return (
+            sabnzbd.cfg.direct_write.get()
+            and self.__cache_limit
+            and (self.__cache_size > self.__cache_size_upper or sabnzbd.Downloader.paused and self.__cache_size)
+            and time.monotonic() > self.__next_flush
+        )
+
+    def run(self):
+        assembler = sabnzbd.Assembler.process
+        nzfs: set[NzbFile] = set()
+
+        while True:
+            article = None
+            nzfs.clear()
+            with self.__cache_size_cv:
+                self.__cache_size_cv.wait_for(
+                    lambda: self.shutdown or self.__needs_flush(),
+                    timeout=5.0,
+                )
+            if self.shutdown:
+                break
+
+            # Timeout so is reach when paused and no further articles arrive
+            if not self.__needs_flush():
+                continue
+
+            self.__next_flush = time.monotonic() + _SECONDS_BETWEEN_FLUSHES
+            for article in self.__article_table.copy():
+                if article.nzf.type == "yenc" and article.nzf not in nzfs and article.nzf.prepare_filepath():
```

Comment:
Is yenc safe enough? Isn't it only if the offset is defined that we can trigger this? Or am I maybe misunderstood what it's checking here?

Reply:
The decoder currently normalise data_begin to 0-indexed, and doesn't include data_end but I could change that.
Currently from `=ypart begin=15360001 end=15744000` the response gives us `part_begin = 15360000` and `part_size = 384000`
I could make it return `part_begin=15360001` `part_end=15744000` and `part_size = 384000`

`=ypart begin=1 end=100000` would give us `part_begin=0` so no way of determining if yenc was missing the offset or if it was non-multipart.

Then we can validate the offsets `not part_begin or not part_end` and revert to append mode.

Probably also needs some guards that offset is less than article.file_size and before any sparse call.

Reply:
If begin and size are defined, that would be same right? So er already have the check